lib/Table/TableStyle.ts

```diff
@@ -397,10 +443,176 @@ export default class TableStyle {
+          data: startDate
+        });
+      }
+    );
+    return intervals;
+  }
+
+  /**
+   * Returns a finish date for each row in the table.
+   */
+  @computed
+  private get finishJulianDates(): (JulianDate | null)[] | undefined {
+    if (this.endTimeColumn) {
+      return this.endTimeColumn.valuesAsJulianDates.values;
+    }
+
+    const timeColumn = this.timeColumn;
+    if (timeColumn === undefined) {
+      return;
+    }
+
+    // If id columns is not defined, group columns by (lat, lon) so that the
+    // finish date for a row with a certain location will be the date for
+    // the next row at the same location.
+    const groupByCols =
+      this.idColumns ||
+      filterOutUndefined([this.latitudeColumn, this.longitudeColumn]);
+    const tableRowIds = this.tableModel.rowIds;
+    const rowGroups = Object.values(
+      groupBy(tableRowIds, rowId =>
+        groupByCols.map(col => col.values[rowId]).join("-")
+      )
+    );
+
+    // Estimate a final duration value to calculate the end date for groups
+    // that have only one row. Fallback to a global default if an estimate
+    // cannot be found.
+    let finalDurationSeconds = DEFAULT_FINAL_DURATION_SECONDS;
+    for (let i = 0; i < rowGroups.length; i++) {
+      const rowIds = rowGroups[i];
+      const startDates = rowIds.map(
+        id => timeColumn.valuesAsJulianDates.values[id]
+      );
+      const sortedStartDates = sortedUniqueDates(startDates);
+      const finalDuration = estimateFinalDurationSeconds(sortedStartDates);
+      if (finalDuration) {
+        finalDurationSeconds = finalDuration;
+        break;
+      }
+    }
+
+    const startDates = timeColumn.valuesAsJulianDates.values;
+    const finishDates: (JulianDate | null)[] = [];
+    rowGroups.forEach(rowIds => {
+      const startDatesForGroup = rowIds.map(id => startDates[id]);
+      const finishDatesForGroup = this.calculateFinishDatesFromStartDates(
+        startDatesForGroup,
+        finalDurationSeconds
+      );
+      finishDatesForGroup.forEach((date, i) => {
+        finishDates[rowIds[i]] = date;
+      });
+    });
+    return finishDates;
+  }
+
+  /**
+   * Computes an and end date for each given start date. The end date for a
+   * given start date is the next higher date in the input. To compute the end
+   * date for the last date in the input, we estimate a duration based on the
+   * average interval between dates in the input. If the input has only one
+   * date, then an estimate cannot be made, in that case we use the
+   * `defaultFinalDurationSeconds` to compute the end date.
+   */
+  private calculateFinishDatesFromStartDates(
+    startDates: (JulianDate | null | undefined)[],
+    defaultFinalDurationSeconds: number
+  ) {
+    const sortedStartDates: JulianDate[] = sortedUniqueDates(startDates);
+    return startDates.map(date => {
+      if (!date) {
+        return null;
+      }
+
+      const nextDateIndex = sortedIndex(sortedStartDates, date);
```

Comment:
Well, I did some testing of it and it seems to work and I'm not sure how.

Reply:
That one is actually a [cesium function](https://cesium.com/docs/cesiumjs-ref-doc/binarySearch.html) that accepts a comparator.

Reply:
Well it works because, the array we are searching in and the array we are iterating are the same array. So it is a simple "is this the same object" comparison. We will need a `JulianDate.compare` if they were two different date objects.

Reply:
Fixed!